src/backends/backend/tensorflow/tensorflow.cc

```diff
@@ -0,0 +1,1923 @@
+    }
+    TRTISTF_ModelMakeCallable(
+        instance->trtistf_model_.get(), input_names.data(), input_types.data(),
+        config_inputs.ArraySize(), output_names.data(), output_types.data(),
+        config_outputs.ArraySize());
+  }
+
+  available_instances_.Push(instance);
+  return nullptr;  // success
+}
+
+ModelState::ModelState(
+    TRITONBACKEND_Model* triton_model, const std::string& name,
+    ni::TritonJson::Value&& model_config)
+    : triton_model_(triton_model), name_(name),
+      model_config_(std::move(model_config)), inflight_thread_count_(0)
+{
+}
+
+ModelState::~ModelState()
+{
+  // Wait for all threads to exit...
+  while (inflight_thread_count_ > 0) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+}
+
+TRITONSERVER_Error*
+ModelState::ValidateModelConfig()
+{
+  // We have the json DOM for the model configuration...
+  ni::TritonJson::WriteBuffer buffer;
+  RETURN_IF_ERROR(model_config_.PrettyWrite(&buffer));
+  TRITONSERVER_LogMessage(
+      TRITONSERVER_LOG_INFO, __FILE__, __LINE__,
+      (std::string("model configuration:\n") + buffer.Contents()).c_str());
+
+  ni::TritonJson::Value ios;
+  RETURN_IF_ERROR(model_config_.MemberAsArray("input", &ios));
+  for (size_t i = 0; i < ios.ArraySize(); i++) {
+    ni::TritonJson::Value io;
+    RETURN_IF_ERROR(ios.IndexAsObject(i, &io));
+    std::string io_name;
+    RETURN_IF_ERROR(io.MemberAsString("name", &io_name));
+    // Check datatypes
+    std::string io_dtype;
+    RETURN_IF_ERROR(io.MemberAsString("data_type", &io_dtype));
+    RETURN_ERROR_IF_TRUE(
+        nib::ConvertDataType(io_dtype) ==
+            TRTISTF_DataType::TRTISTF_TYPE_INVALID,
+        TRITONSERVER_ERROR_INVALID_ARG,
+        std::string("unsupported datatype '") + io_dtype + "' for tensor '" +
+            io_name + "' for model '" + name_ + "'");
+  }
+  RETURN_IF_ERROR(model_config_.MemberAsArray("output", &ios));
+  for (size_t i = 0; i < ios.ArraySize(); i++) {
+    ni::TritonJson::Value io;
+    RETURN_IF_ERROR(ios.IndexAsObject(i, &io));
+    std::string io_name;
+    RETURN_IF_ERROR(io.MemberAsString("name", &io_name));
+    // Check datatypes
+    std::string io_dtype;
+    RETURN_IF_ERROR(io.MemberAsString("data_type", &io_dtype));
+    RETURN_ERROR_IF_TRUE(
+        nib::ConvertDataType(io_dtype) ==
+            TRTISTF_DataType::TRTISTF_TYPE_INVALID,
+        TRITONSERVER_ERROR_INVALID_ARG,
+        std::string("unsupported datatype '") + io_dtype + "' for tensor '" +
+            io_name + "' for model '" + name_ + "'");
+  }
+
+  return nullptr;  // success
+}
+
+void
+ModelState::ProcessRequest(
+    TRITONBACKEND_Request** requests, const uint32_t request_count,
+    const uint64_t exec_start_ns)
+{
+  inflight_thread_count_++;
+  auto instance = available_instances_.Pop();
+
+  // Currently launch thread for each batch, but we may launch long-running
+  // thread for each instance.
+  std::thread process_thread(
```

Comment:
I think performance-wise this will be an issue, but we can revisit that later. Also will likely need to run using a specific instance due to rate-limiter changes.

Reply:
I ran perf_client (on `graphdef_float32_float32_float32`) and it did show performance-wise issue, changed to long-running instance thread helps but it still shows that is less performant than not using backend API.